storagenode/gracefulexit/worker.go

```diff
@@ -4,48 +4,248 @@
 package gracefulexit
 
 import (
+	"bytes"
 	"context"
+	"io"
+	"os"
 	"time"
 
 	"github.com/zeebo/errs"
 	"go.uber.org/zap"
 
+	"storj.io/storj/pkg/pb"
+	"storj.io/storj/pkg/rpc"
 	"storj.io/storj/pkg/storj"
+	"storj.io/storj/storagenode/pieces"
 	"storj.io/storj/storagenode/satellites"
+	"storj.io/storj/storagenode/trust"
+	"storj.io/storj/uplink/ecclient"
 )
 
 // Worker is responsible for completing the graceful exit for a given satellite.
 type Worker struct {
 	log         *zap.Logger
-	satelliteID storj.NodeID
+	store       *pieces.Store
 	satelliteDB satellites.DB
+	trust       *trust.Pool
+	dialer      rpc.Dialer
+	satelliteID storj.NodeID
+	ecclient    ecclient.Client
 }
 
 // NewWorker instantiates Worker.
-func NewWorker(log *zap.Logger, satelliteDB satellites.DB, satelliteID storj.NodeID) *Worker {
+func NewWorker(log *zap.Logger, store *pieces.Store, satelliteDB satellites.DB, trust *trust.Pool, dialer rpc.Dialer, satelliteID storj.NodeID) *Worker {
 	return &Worker{
 		log:         log,
-		satelliteID: satelliteID,
+		store:       store,
 		satelliteDB: satelliteDB,
+		trust:       trust,
+		dialer:      dialer,
+		satelliteID: satelliteID,
+		ecclient:    ecclient.NewClient(log, dialer, 0),
 	}
 }
 
 // Run calls the satellite endpoint, transfers pieces, validates, and responds with success or failure.
 // It also marks the satellite finished once all the pieces have been transferred
+// TODO handle transfers in parallel
 func (worker *Worker) Run(ctx context.Context, satelliteID storj.NodeID, done func()) (err error) {
 	defer mon.Task()(&ctx)(&err)
 	defer done()
+
 	worker.log.Debug("running worker")
 
-	// TODO actually process the order limits
-	// https://storjlabs.atlassian.net/browse/V3-2613
+	addr, err := worker.trust.GetAddress(ctx, satelliteID)
+	if err != nil {
+		return errs.Wrap(err)
+	}
+
+	conn, err := worker.dialer.DialAddressID(ctx, addr, satelliteID)
+	if err != nil {
+		return errs.Wrap(err)
+	}
+	defer func() {
+		err = errs.Combine(err, conn.Close())
+	}()
+
+	client := conn.SatelliteGracefulExitClient()
+
+	c, err := client.Process(ctx)
+	if err != nil {
+		return errs.Wrap(err)
+	}
+
+	for {
+		response, err := c.Recv()
+		if errs.Is(err, io.EOF) {
+			// Done
+			break
+		}
+		if err != nil {
+			// TODO what happened
+			return errs.Wrap(err)
+		}
+
+		switch msg := response.GetMessage().(type) {
+		case *pb.SatelliteMessage_NotReady:
+			break // wait until next worker execution
+		case *pb.SatelliteMessage_TransferPiece:
+			pieceID := msg.TransferPiece.OriginalPieceId
+			reader, err := worker.store.Reader(ctx, satelliteID, pieceID)
+			if err != nil {
+				transferErr := pb.TransferFailed_UNKNOWN
+				if errs.Is(err, os.ErrNotExist) {
+					transferErr = pb.TransferFailed_NOT_FOUND
+				}
+				worker.log.Error("failed to get piece reader.", zap.String("satellite ID", satelliteID.String()), zap.String("piece ID", pieceID.String()), zap.Error(errs.Wrap(err)))
```

Comment:
nit - I think use zap.Stringer instead of zap.String - then you can just pass in satelliteID instead of satelliteID.String()

Reply:
Updated